### PR TITLE
feat: get all files by their exports are unused

### PR DESCRIPTION
**Pull request by brandonxiang**
#253

### CODE REVIEW

File: src/analyzer.ts [hunk 1]
```diff
@@ -244,4 +246,47 @@ export default (
...
+  });
+
+  return {
+    unusedFiles,
+    unusedExports: res,
+  };
+}
+
+function arrayEqual(a: string[], b: string[]) {
+  for (let i = 0; i < a.length; i++) {
```
Comment (mrseanryan) [inline src/analyzer.ts:286]: With `forEach()` could be shorter:

a.forEach(aItem =>

**Comment by mrseanryan** [comment]
Thank you for the PR.

Can you please add an itest to cover.

This involves adding another example under ./example
and then adding a call in:

ispec/run.sh

---

Also, please add an entry at top of CHANGELOG.md

**Comment by mrseanryan** [comment]
Also to `show the output` - need to extend `cli.ts`

Unless the `silent` option is on, then we should display the "completely unused files".

Those can be show, after the existing messages.

**Comment by mrseanryan** [comment]
TODO - fix so the itests pass

(should not need to change the itests)

Probably this can be fixed by updating `cli.ts`

**Review by mrseanryan — CHANGES_REQUESTED** [review]
see comments, thank you

**Comment by brandonxiang** [comment]
OK, I will spend some time to refactor that code.

**Comment by mrseanryan** [comment]
good, I will look again when you push

**Comment by brandonxiang** [comment]
image
I have a confused whether I should put the completelyUnused attribute inside the Analysis Object. 

Whether I should use the original analysis function?

**Comment by mrseanryan** [comment]
> I have a confused whether I should put the completelyUnused attribute inside the Analysis Object.
Whether I should use the original analysis function?

Let's add it to the Analysis type, as optional. Like:

```
export interface Analysis {
  [index: string]: ExportNameAndLocation[];
  unusedFiles?: string[]; // <<<<<<
}
```

Then, we can still just export 1 function, as before:

```
export default (
  files: File[],
  extraOptions?: ExtraCommandLineOptions,
): Analysis => {
 // (existing code)
 // ...

 if (extraOptions.findCompletelyUnusedFiles) // <<<
   analysis.unusedFiles = getUnusedFiles(files, extraOptions);  // <<<

return analysis;
}
```

This extends the API rather than changes it, which is less risky and is easier for users to pick up...

**Comment by mrseanryan** [comment]
TODO:

* [ ] add an option `findCompletelyUnusedFiles` and doc it in the README

**Comment by brandonxiang** [comment]
1. I have put the unusedFiles logic into the analysis function. All calculation is only only.
2. I have written the itest by cucumber-js and finished the doc

**Comment by mrseanryan** [comment]
* [ ] add CHANGELOG entry

**Comment by mrseanryan** [comment]
hi brandonxiang

There were a couple of things I wanted to tidy up (and also avoid adding lodash)
so I just made a new PR.

It is merged!

Will release shortly, as new Major version.

Thank you for contributing !

**Pull request closed without merging**